src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/sequences/aggregate/CountFunctionIterator.java

```diff
@@ -50,36 +49,36 @@ public Item next() {
         if (this._hasNext) {
             RuntimeIterator iterator = this._children.get(0);
             if (!iterator.isRDD()) {
-                if (_children.get(0) instanceof VariableReferenceIterator) {
-                    VariableReferenceIterator expr = (VariableReferenceIterator) _children.get(0);
+                if (iterator instanceof VariableReferenceIterator) {
+                    VariableReferenceIterator expr = (VariableReferenceIterator) iterator;
                     this._hasNext = false;
-                    return _currentDynamicContext.getVariableCount(expr.getVariableName());
+                    return _currentDynamicContext.getLocalVariableCount(expr.getVariableName());
                 }
                 List<Item> results = getItemsFromIteratorWithCurrentContext(iterator);
                 this._hasNext = false;
                 return ItemFactory.getInstance().createIntegerItem(results.size());
             } else {
-                Long count = iterator.getRDD(_currentDynamicContext).count();
+                long count = iterator.getRDD(_currentDynamicContext).count();
                 this._hasNext = false;
                 if (count > (long) Integer.MAX_VALUE) {
                     // TODO: handle too big x values
                     throw new SparksoniqRuntimeException("The count value is too big to convert to integer type.");
                 } else {
-                    return ItemFactory.getInstance().createIntegerItem(count.intValue());
+                    return ItemFactory.getInstance().createIntegerItem((int) count);
```

Comment:
A side note for later: we may want at some point to make integer items 64-bit rather than 32, so as to avoid these conversions.

Reply:
Issue created at #385